### ISSUE TITLE
perf(GPU): Mixed-precision training (TF32 / FP16 / BF16) is not engaged on the standard Train path — fp32 models pay full fp32 compute cost on TF32-capable hardware

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#337]
## Problem

On Ampere+ NVIDIA hardware (RTX 30/40/50 series, A100, H100), cuBLAS supports TF32 matmul accumulation for fp32 inputs: the multiply is rounded to a 10-bit mantissa and the accumulate stays at full fp32, yielding ~5× the FLOPs of strict fp32 with bit-for-bit-comparable training-loss curves on every published benchmark.

`CudaBackend` does not call `cublasSetMathMode(handle, CUBLAS_TF32_TENSOR_OP_MATH)`. fp32 matmuls run at full fp32 cost. Users on Ampere+ are leaving 5× perf on the table on every GEMM-heavy model (BERT, ViT, Stable Diffusion's UNet projection layers).

Similarly:
- FP16 / BF16 element types exist as types (`System.Half`, `AiDotNet.Tensors.NumericOperations.BFloat16`) and the type table in `WeightRegistry.StreamableTypeSize` registers them, but the GPU kernel cache has no `_half` / `_bf16` variants for Conv2D, MatMul, BatchNorm. A model declared as `Network<Half>` falls all the way through to scalar fp32 emulation on GPU.
- The standard Train path on the AiDotNet consumer side hard-codes `T` from the model's generic parameter, so even "use fp16" requires the user to rebuild the entire network as `VGGNetwork<Half>` — there's no automatic mixed-precision (autocast) where the framework picks fp16 for compute-bound layers and keeps fp32 for the master weights.

### 1. TF32 by default on Ampere+ for fp32

In `CudaBackend.Initialize`, after creating the cuBLAS handle:

```csharp
if (deviceCapabilityMajor >= 8 && CudaDispatchPolicy.AllowTF32 /* default true */)
    CuBlasNative.cublasSetMathMode(_cublasHandle, CUBLAS_TF32_TENSOR_OP_MATH);
```

Plus an opt-out via env var `AIDOTNET_DISABLE_TF32` and a `CudaDispatchPolicy.AllowTF32` property for code-level override (numerical reproducibility runs).

### 2. FP16 / BF16 kernel variants

Add to the kernel cache:
- `gemm_half`, `gemm_bfloat16` (cuBLAS HGEMM / cublasGemmEx with appropriate compute types)
- `conv2d_half`, `conv2d_bfloat16` (cuDNN with appropriate `CUDNN_DATA_HALF` / `CUDNN_DATA_BFLOAT16` descriptors)
- `batchnorm_half`, `batchnorm_bfloat16` (cuDNN persistent BN with the corresponding compute type)
- `softmax_half`, `softmax_bfloat16`

These cover the GEMM-and-conv-bound 80 % of training time on transformer / vision models.

### 3. Autocast (automatic mixed precision)

A `MixedPrecisionPolicy` on `AiDotNetEngine` and per-layer:
- `Layers default to fp32 master weights.`
- Conv/MatMul/Attention dispatch in fp16 or bf16 when compute is bound (`PolicyCompute = Half` or `BFloat16`); accumulate in fp32 via cuBLAS's compute-type parameter.
- Loss scaling for fp16 (BF16 doesn't need it — same exponent range as fp32).
- Gradient unscale before optimizer step.

API surface:
```csharp
network.WithAutoMixedPrecision(MixedPrecisionPolicy.FP16);  // or BFloat16, or Disabled
```

## What this buys

- TF32 on Ampere: ~5× cuBLAS GEMM throughput for fp32 models on RTX 30/40, A100, H100 with no accuracy cost.
- BF16 autocast: ~2× of TF32 on Hopper / Ada with negligible accuracy cost for transformer training.
- FP16 autocast: same throughput as BF16 on Ampere (which doesn't have native BF16 cores for some ops), but requires loss scaling.

## What is NOT in scope

- FP8 (Hopper E4M3/E5M2) — separate effort.
- INT8 training quantization (QAT) — separate effort. INT8 inference quantization tracked under #209.

## Tracking

Implementation lands in PR ooples/AiDotNet.Tensors#333.
